### ISSUE TITLE
transforming data?

### ISSUE THREAD

**Issue by nikkisingh111333** [oracle/tribuo#100]
hey i want a good tutorial on data transformation.i m following docs but getting confused by similar terminology like transformerMap transform transformation transformationMap etc..can you point me out to the quick start of transformation of dataset ..so that i can easily apply normalization binning vectorization all kind of data processesing stuff?

**Comment by Craigacp** [comment]
Vectorization is a different matter, Tribuo has text processing pipelines in `org.tribuo.data.text` for vectorising text data, and once it's been vectorised it can be transformed using the transformation machinery. There are example text pipelines in the [columnar tutorial](https://tribuo.org/learn/4.0/tutorials/columnar-tribuo-v4.html) which extract unigrams and bigrams. You can do much more complicated things with it if you want, for example we've got things that produce word vectors internally based on our in house word embeddings implementation.

As for the transform package, we'll add a transformation tutorial to the todo list. In the meantime I'll cover the core concepts below.

`org.tribuo.transform` has two main parts: Transformations and Transformers (yes the naming isn't the best, we're open to suggestions for new names for the next major version bump). A Transformation represents the process that you want to apply to a feature, but hasn't been fit to the data yet. A Transformer is the transformation after it's been fit to the data. Similarly to the rest of Tribuo, we use the type system to separate out trained objects from untrained ones. You can think of a Tribuo Transformation as a scikit-learn Transformer before fit has been called, and a Tribuo Transformer is a scikit-learn Transformer after fit has been called. Transformations/Transformers operate independently on each feature, they do not share information across features (so you can't use them to do PCA, or conditional transformations based on the values of other features). We're considering how to integrate PCA and other feature extraction algorithms into the transformation framework, but there's nothing public yet.

A `TransformationMap` is a set of lists of `Transformation`s to apply, one per feature. It also has "global transformations" if you want to apply the same transformation to every feature, and a way of applying a list of transformations to any feature which matches the specified regex. It produces a `TransformerMap` when fit to a dataset with `Dataset.createTransformers(TransformationMap)`.

A `TransformerMap` is a set of lists of fitted `Transformer`s which can transform a dataset or example in a repeatable way. The `transformDataset` and `transformExample` methods produce a copy of the input object, or you can use `MutableDataset.transform(TransformerMap)` to apply it in place to a specific dataset. This is used to transform a test set so it's processed the same way as the training set was.

As this is a lot of stuff to deal with, we also provide `TransformTrainer` which accepts a `TransformationMap` and a `Trainer`, then when `TransformTrainer.train` is called it first fits the `TransformationMap` on the supplied dataset to produce a `TransformerMap`, then it transforms the training dataset (producing a copy of it), then trains a model using the supplied trainer, then the model is wrapped in `TransformedModel` which automatically applies the transfomers to any incoming data (to ensure that the test time data is transformed in the same way as the training time data). Basically `TransformTrainer` does what you want it to do, wrapping up the transformations, fitting them to the training time data, then applying them to the training data and any test time data the model sees.

Tribuo records all these transformations in the model & dataset provenances, so they are recoverable from a model itself. You can see this in the [configuration tutorial](https://tribuo.org/learn/4.0/tutorials/configuration-tribuo-v4.html#Feature-Transformations).

There is one further thing which increases the complexity of the system, Tribuo's sparsity assumption. In general transformers and transformations do not operate on the implicit zeros present in Tribuo's examples and datasets. If you do want them to operate on those implicit zeros then you should call `MutableDataset.densify` to make them into explicit zeros before applying any transformations. `TransformerMap` and `TransformTrainer` have a `densify` argument/field in the appropriate places to allow the user to specify this if they don't want to densify the dataset first. However you should know that this interaction is tricky, and you should be careful when using `TransformerMap` directly to ensure that densify has been set consistently between when the transformations were fitted and when the transformers are applied.

**Comment by nikkisingh111333** [comment]
okk okkk but how can i predict high dimensional unigram/bigram featured
data. i mean should i directly pass text column data like 'i want to party'
instead of makking field of each word.['i','want','to','party'] then append
it to Immutablemap.of() function for predictions. ?

**Comment by Craigacp** [comment]
You can generate unigrams and bigrams using a `TextPipeline`. As I mentioned above there is a simple example of unigram and bigram processing in the columnar tutorial.

For example to use a pipeline outside of a datasource you can do:
```java
var pipeline = new BasicPipeline(new BreakIteratorTokenizer(Locale.US),2);
var tfe = new TextFeatureExtractor(pipeline);
```

will produce a text pipeline that creates examples based on input text by generating unigrams and bigrams and returning a binary feature for each unigram or bigram (i.e. it's not term counting). If you want term counting you can use `TokenPipeline` which can do term counting and optionally feature hashing into a specific hash size.

The tutorial shows how to do that as part of a columnar data source and process the text in a specific field. Or you can use a `TextDataSource` sublclass to load in text data in a one document per line format. It all depends where your data is and how you want to load it.

**Comment by nikkisingh111333** [comment]
Cool, thanks! i have implemented it successfully but everytime i predict
most of the time i get same answer .i m practicing xgboost with spam/ham
email detection and most of the time if get spam even if its not but my
test score is great 95% with both higher precision/recall and good f1
score.why is that ? i m using token-pipeline with 2 ngrams.?

**Comment by Craigacp** [comment]
I'm not clear what the issue is. You say that most of the time it's predicting spam, but the test accuracy is high. Is the test dataset imbalanced? The default toString output of the evaluation will print the number of each class which should help you run this down.

**Comment by nikkisingh111333** [comment]
Yes, it is. its a imbalance dataset and i dont know how to make it
balanced.i know two algorithm smote and random data generation .or may be
tribuo has its own way to deal with imbalance dataset which i dont know ?
is their anything for me?

**Comment by Craigacp** [comment]
We discussed imbalanced data in the other issue, and what I said there still stands. My question was if the performance metrics showed an issue. Tribuo computes the per class f1 which you can use to determine if it's correctly predicting each class separately or if the imbalanced data has actually caused an issue with the classifier's ability to learn each class. If you're getting good per class f1s then it's probably solved the problem fairly well. Otherwise it's very data dependent to figure out what the issue is, and that's a machine learning problem rather than an issue with Tribuo.

**Comment by nikkisingh111333** [comment]
Ok, thanks for the clarification.one more question before i ll wrap this
issue up .IdentityProcessor turns  classes into features using one-hot
encoding.but thats not my requirement.i want single feature with unique
numbers as for classififcation.like this
["cat","dog","cat","lizard","dog"]=[1,2,1,3,2] rather than separate
features for each class how can i do this i was digging docs but found
nothing.?

**Comment by Craigacp** [comment]
To convert a set of categoricals into a single feature with different values you'll need to write your own `FieldProcessor`. Be careful as they are supposed to be stateless, but if you want the set of categoricals to be inferred from the data then it will make the `FieldProcessor` stateful, which will cause issues with reinstantiating it from the provenance and using it a test time. I recommend you make a `FieldProcessor` which accepts a `Map<String,Integer>` on construction, as that will be properly recorded and rebuilt.

**Comment by nikkisingh111333** [comment]
Thanks for the tip! so what necesary interfaces or class i needed for
it.any hint?

**Comment by Craigacp** [comment]
You'll need to implement the `org.tribuo.data.columnar.FieldProcessor` interface.

**Comment by nikkisingh111333** [comment]
Thanks, I'll check it out. thanks for your support.

**Comment by nikkisingh111333** [comment]
thanks for your support  i ll check it out.
